File: android/app/src/androidTest/kotlin/dev/omnesis/android/journeys/JourneyGateway.kt
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

package dev.omnesis.android.journeys

import androidx.test.platform.app.InstrumentationRegistry
import dev.omnesis.android.transport.tls.LeafCertPinner
import java.net.URI
import java.util.concurrent.TimeUnit
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

/**
 * The synthetic gateway a journey drives the app against.
 *
 * `scripts/run-mobile-journeys.sh android` boots it on the host and passes its
 * address as the emulator reaches it (`10.0.2.2`), its admin token and its TLS
 * leaf fingerprint as instrumentation arguments. The token belongs to that
 * throwaway gateway only. The helper talks to the gateway directly to set a
 * journey up — minting a pairing code, reading back what the app did — and
 * never stands in for the app itself.
 */
class JourneyGateway private constructor(
    val url: String,
    private val token: String,
    val fingerprint: String,
) {
    private val pinner = LeafCertPinner(fingerprint)
    private val http = OkHttpClient.Builder()
        .sslSocketFactory(pinner.socketFactory, pinner.trustManager)
        .hostnameVerifier(pinner.hostnameVerifier)
        .callTimeout(45, TimeUnit.SECONDS)
        .build()

    /** `host:port` as the pairing confirmation sheet shows it. */
    val hostAndPort: String
        get() = URI(url).let { "${it.host}:${it.port}" }

    /** Mints a one-time Android pairing code and returns it with the QR payload the gateway would show. */
    fun mintPairing(deviceName: String): Pairing {
        val pending = post(
            "/admin/devices/pair",
            buildJsonObject {
                put("kind", "android")
                put("name", deviceName)
            },
        )
        val code = pending.string("pairingCode")
        val qr = post(
            "/admin/devices/pair-qr",
            buildJsonObject {
                put("pairingCode", code)
                put("gatewayUrl", url)
                put("trustMode", "pinned-leaf")
            },
        )
        return Pairing(code = code, payload = qr.string("qrPayload"))
    }

    /** Whether the gateway lists a device of [kind] with this name. */
    fun hasDevice(name: String, kind: String): Boolean {
        val items: JsonArray = request(Request.Builder().url("$url/admin/devices").get())["items"]?.jsonArray
            ?: return false
        return items.any { item ->
            val device = item.jsonObject
            device["name"]?.jsonPrimitive?.content == name && device["kind"]?.jsonPrimitive?.content == kind
        }
    }

    data class Pairing(val code: String, val payload: String)

    private fun post(path: String, body: JsonObject): JsonObject =
        request(Request.Builder().url("$url$path").post(body.toString().toRequestBody(JSON)))

    private fun request(builder: Request.Builder): JsonObject {
        val request = builder.header("Authorization", "Bearer $token").build()
        http.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            check(response.isSuccessful) { "The gateway answered ${request.url.encodedPath} with HTTP ${response.code}: $text" }
            return Json.parseToJsonElement(text).jsonObject
        }
    }

    private fun JsonObject.string(key: String): String =
        checkNotNull(this[key]?.jsonPrimitive?.content) { "The gateway's answer has no $key" }

    companion object {
        private val JSON = "application/json".toMediaType()

        /**
         * Reads the instrumentation arguments the journey script supplies. A run
         * without them fails rather than skipping: a silent skip would read as a pass.
         */
        fun fromInstrumentationArguments(): JourneyGateway {
            val args = InstrumentationRegistry.getArguments()
            fun required(name: String): String = checkNotNull(args.getString(name)?.takeIf(String::isNotBlank)) {
                "Instrumentation argument $name is missing. Run the journeys through scripts/run-mobile-journeys.sh android."
            }
            return JourneyGateway(
                url = required("journeyGatewayUrl").trimEnd('/'),
                token = required("journeyToken"),
                fingerprint = required("journeyFingerprint").lowercase(),
            )
        }
    }
}

File: ios/UITests/OmnesisDemoUITests/JourneyGateway.swift
// SPDX-License-Identifier: AGPL-3.0-or-later
// Copyright (c) 2026 Adrien Conrath

import CryptoKit
import Foundation
import XCTest

/// The synthetic gateway a journey drives the app against.
///
/// `scripts/run-mobile-journeys.sh ios` boots it and hands its address, admin
/// token and TLS leaf fingerprint to the on-simulator test runner through
/// `TEST_RUNNER_OMNESIS_JOURNEY_CONFIG` (xcodebuild strips the prefix). The
/// token only ever belongs to that throwaway gateway. The helper talks to the
/// gateway directly to set a journey up — minting a pairing code, reading back
/// what the app did — and never stands in for the app itself.
struct JourneyGateway {
    let url: URL
    let token: String
    let fingerprint: String

    private struct Config: Decodable {
        let gatewayURL: String
        let token: String
        let fingerprint: String
    }

    /// Reads the configuration the journey script supplies. A journey run
    /// without it fails rather than skipping: the journeys exist to be run
    /// against a gateway, and a silent skip would read as a pass.
    static func fromEnvironment() throws -> JourneyGateway {
        guard let raw = ProcessInfo.processInfo.environment["OMNESIS_JOURNEY_CONFIG"],
              let data = raw.data(using: .utf8)
        else {
            throw JourneyGatewayError.missingConfiguration
        }
        let config = try JSONDecoder().decode(Config.self, from: data)
        guard let url = URL(string: config.gatewayURL) else {
            throw JourneyGatewayError.invalidConfiguration(config.gatewayURL)
        }
        return JourneyGateway(url: url, token: config.token, fingerprint: config.fingerprint.lowercased())
    }

    /// `host:port` as the pairing confirmation sheet shows it.
    var hostAndPort: String {
        "\(url.host ?? ""):\(url.port ?? 443)"
    }

    /// The launch-environment pairing the DEBUG app seeds into its Keychain
    /// (`DEMO_PAIRING_JSON`), for journeys that start from a paired app.
    var automationPairingJSON: String {
        let object: [String: String] = [
            "url": url.absoluteString,
            "token": token,
            "fingerprint": fingerprint,
            "name": "Journey Gateway",
        ]
        let data = (try? JSONSerialization.data(withJSONObject: object, options: [.sortedKeys])) ?? Data()
        return String(data: data, encoding: .utf8) ?? "{}"
    }

    /// Mints a one-time iPhone pairing code and returns the QR payload the
    /// gateway would show for it, pinned to the gateway's own certificate.
    func mintPairingPayload(deviceName: String) throws -> String {
        let pending = try post("/admin/devices/pair", body: ["kind": "ios", "name": deviceName])
        guard let code = pending["pairingCode"] as? String else {
            throw JourneyGatewayError.unexpectedResponse("/admin/devices/pair")
        }
        let qr = try post(
            "/admin/devices/pair-qr",
            body: ["pairingCode": code, "gatewayUrl": url.absoluteString, "trustMode": "pinned-leaf"]
        )
        guard let payload = qr["qrPayload"] as? String else {
            throw JourneyGatewayError.unexpectedResponse("/admin/devices/pair-qr")
        }
        return payload
    }

    /// Whether the gateway lists a live device of `kind` with this name.
    func hasDevice(named name: String, kind: String) throws -> Bool {
        let page = try request("GET", "/admin/devices", body: nil)
        let items = page["items"] as? [[String: Any]] ?? []
        return items.contains { item in
            item["name"] as? String == name && item["kind"] as? String == kind
        }
    }

    private func post(_ path: String, body: [String: String]) throws -> [String: Any] {
        try request("POST", path, body: body)
    }

    private func request(_ method: String, _ path: String, body: [String: String]?) throws -> [String: Any] {
        var request = URLRequest(url: url.appendingPathComponent(path))
        request.httpMethod = method
        request.timeoutInterval = 30
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let session = URLSession(
            configuration: .ephemeral,
            delegate: PinnedLeafDelegate(fingerprint: fingerprint),
            delegateQueue: nil
        )
        defer { session.finishTasksAndInvalidate() }

        var outcome: Result<(Data, URLResponse), Error>?
        let done = DispatchSemaphore(value: 0)
        session.dataTask(with: request) { data, response, error in
            if let data, let response {
                outcome = .success((data, response))
            } else {
                outcome = .failure(error ?? JourneyGatewayError.unexpectedResponse(path))
            }
            done.signal()
        }.resume()
        guard done.wait(timeout: .now() + 45) == .success, let outcome else {
            throw JourneyGatewayError.timedOut(path)
        }
        let (data, response) = try outcome.get()
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200 ..< 300).contains(status) else {
            let text = String(data: data, encoding: .utf8) ?? ""
            throw JourneyGatewayError.httpStatus(path, status, text)
        }
        return (try JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
    }
}

enum JourneyGatewayError: Error, CustomStringConvertible {
    case missingConfiguration
    case invalidConfiguration(String)
    case unexpectedResponse(String)
    case timedOut(String)
    case httpStatus(String, Int, String)

    var description: String {
        switch self {
        case .missingConfiguration:
            "OMNESIS_JOURNEY_CONFIG is not set. Run the journeys through scripts/run-mobile-journeys.sh ios."
        case .invalidConfiguration(let url):
            "The journey configuration names an invalid gateway URL: \(url)"
        case .unexpectedResponse(let path):
            "The gateway answered \(path) without the expected fields."
        case .timedOut(let path):
            "The gateway did not answer \(path) in time."
        case .httpStatus(let path, let status, let body):
            "The gateway answered \(path) with HTTP \(status): \(body)"
        }
    }
}

/// Trusts exactly the gateway's self-signed leaf certificate, the way a
/// paired app does, so the setup calls never disable certificate checks.
private final class PinnedLeafDelegate: NSObject, URLSessionDelegate {
    private let fingerprint: String

    init(fingerprint: String) {
        self.fingerprint = fingerprint
    }

    func urlSession(
        _ session: URLSession,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
    ) {
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let trust = challenge.protectionSpace.serverTrust,
              let leaf = (SecTrustCopyCertificateChain(trust) as? [SecCertificate])?.first
        else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }
        let digest = SHA256.hash(data: SecCertificateCopyData(leaf) as Data)
            .map { String(format: "%02x", $0) }
            .joined()
        if digest == fingerprint {
            completionHandler(.useCredential, URLCredential(trust: trust))
        } else {
            completionHandler(.cancelAuthenticationChallenge, nil)
        }
    }
}
